### PR TITLE
add ipex moe implementation to support Mixtral and PhiMoe

### PR DESCRIPTION
**Pull request by sywangyi**
@OlivierDehaene OR @Narsil

### CODE REVIEW

**Comment by sywangyi** [comment]
need xpu ipex to upgrade to 2.5 as well. or else it will block tgi-xpu

**Comment by Narsil** [comment]
This is still marked as draft, should we review/merge or not ? The code change looks Ok.

**Comment by sywangyi** [comment]
It also need the upgrade of xpu ipex too. once ready, I will mark it ready for review. thanks

**Comment by sywangyi** [comment]
@Narsil please help review the PR. thanks

**Comment by sywangyi** [comment]
@Narsil  could you review it. the PR also upgrade xpu ipex which means to support AWQ and GPTQ

**Comment by sywangyi** [comment]
@danieldk could you help?

**Comment by danieldk** [comment]
Looking at it now!

File: Dockerfile_intel [hunk 1]
```diff
@@ -83,16 +83,31 @@ RUN wget -qO - https://repositories.intel.com/gpu/intel-graphics.key | gpg --dea
 RUN wget -O- https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS.PUB \
 | gpg --dearmor | tee /usr/share/keyrings/oneapi-archive-keyring.gpg > /dev/null && echo "deb [signed-by=/usr/share/keyrings/oneapi-archive-keyring.gpg] https://apt.repos.intel.com/oneapi all main" | tee /etc/apt/sources.list.d/oneAPI.list
 
-RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt install -y intel-basekit xpu-smi cmake ninja-build pciutils
+RUN wget -O- https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS.PUB \
+  | gpg --dearmor > /tmp/intel-for-pytorch-gpu-dev-keyring.gpg
+
+RUN mv /tmp/intel-for-pytorch-gpu-dev-keyring.gpg /usr/share/keyrings
+
+RUN echo "deb [signed-by=/usr/share/keyrings/intel-for-pytorch-gpu-dev-keyring.gpg] https://apt.repos.intel.com/intel-for-pytorch-gpu-dev all main" > /tmp/intel-for-pytorch-gpu-dev.list
```
Comment (danieldk) [inline Dockerfile_intel:91]:
```suggestion
RUN echo "deb [signed-by=/usr/share/keyrings/oneapi-archive-keyring.gpg] https://apt.repos.intel.com/intel-for-pytorch-gpu-dev all main" > /tmp/intel-for-pytorch-gpu-dev.list
```

Looks like it's the same key as the repo above?

File: server/text_generation_server/layers/moe/__init__.py [hunk 2]
```diff
@@ -29,6 +29,8 @@
     from vllm.model_executor.layers.fused_moe import fused_topk
 elif SYSTEM != "ipex":
     from moe_kernels.fused_moe import fused_topk, grouped_topk
+else:
+    from intel_extension_for_pytorch.llm.modules import GatedMLPMOE
```
Comment (danieldk) [inline server/text_generation_server/layers/moe/__init__.py:33]:
```suggestion
elif SYSTEM == "ipex":
    from intel_extension_for_pytorch.llm.modules import GatedMLPMOE
else:
    from moe_kernels.fused_moe import fused_topk, grouped_topk
```

would be clearer.

File: server/text_generation_server/layers/moe/__init__.py [hunk 3]
```diff
@@ -152,6 +158,17 @@ def forward(self, x: torch.Tensor, *, gating_output: torch.Tensor) -> torch.Tens
...
+            return self.ipex_fused_moe(
+                hidden_states=x,
+                router_logits=gating_output,
+                top_k=self.topk,
+                renormalize=self.renormalize,
+                use_grouped_topk=self.n_expert_group is not None,
+                num_expert_group=self.n_expert_group,
+                topk_group=self.topk_group,
+            )
+
```
Comment (danieldk) [inline server/text_generation_server/layers/moe/__init__.py:171]: I am not sure if we want this here, the idea of this class is to provide the a non-fused implementation. It's a fallback for when no fused implementation is available (e.g. for platforms where no fused kernels are available for GPTQ/AWQ).

**Review by danieldk — COMMENTED** [review]
Thanks, it's nice to have fused MoE on ipex. I added some comments.

File: server/text_generation_server/layers/moe/__init__.py (hunk 3 above)
Comment (sywangyi) [reply server/text_generation_server/layers/moe/__init__.py:171]: from moe_kernels.fused_moe import fused_topk, grouped_topk,  is moe_kernel only for cuda?

Comment (danieldk) [reply server/text_generation_server/layers/moe/__init__.py:171]: Ah, yes, I see the issue now. Thanks!

**Review by danieldk — APPROVED** [review]

**Pull request merged**